load-test-framework/src/main/java/com/google/pubsub/flic/controllers/Client.java

```diff
@@ -0,0 +1,239 @@
+              loadtestLengthSeconds).build());
+    }
+    switch (clientType) {
+      case CPS_GCLOUD_SUBSCRIBER:
+        requestBuilder.setPubsubOptions(PubsubOptions.newBuilder()
+            .setMaxMessagesPerPull(10)
+            .setSubscription(subscription));
+        break;
+      case KAFKA_PUBLISHER:
+      case KAFKA_SUBSCRIBER:
+        requestBuilder.setKafkaOptions(KafkaOptions.newBuilder()
+            .setBroker(broker)
+            .setPollLength(100));
+        break;
+    }
+    StartRequest request = requestBuilder.build();
+    SettableFuture<Void> startFuture = SettableFuture.create();
+    stub = getStub();
+    stub.start(request, new StreamObserver<StartResponse>() {
+      private int connectionErrors = 0;
+      @Override
+      public void onNext(StartResponse response) {
+        log.info("Successfully started client [" + networkAddress + "]");
+        clientStatus = ClientStatus.RUNNING;
+      }
+
+      @Override
+      public void onError(Throwable throwable) {
+        if (connectionErrors > 10) {
+          log.error("Client failed to start " + connectionErrors + " times, shutting down.");
+          clientStatus = ClientStatus.FAILED;
+          startFuture.setException(throwable);
+          doneFuture.setException(throwable);
+          return;
+        }
+        connectionErrors++;
+        try {
+          Thread.sleep(5000);
+        } catch (InterruptedException e) {
+          log.info("Interrupted during back off, retrying.");
+        }
+        log.debug("Going to retry client connection, likely due to start up time.");
+        stub = getStub();
+        stub.start(request, this);
+      }
+
+      @Override
+      public void onCompleted() {
+        startFuture.set(null);
+      }
+    });
+    try {
+      startFuture.get();
+      executorService.scheduleAtFixedRate(this::checkClient, 20, 20, TimeUnit.SECONDS);
+    } catch (ExecutionException e) {
+      throw e.getCause();
+    }
+  }
+
+  private void checkClient() {
+    if (clientStatus != ClientStatus.RUNNING) {
+      return;
+    }
+    stub.check(LoadtestProto.CheckRequest.getDefaultInstance(), new StreamObserver<LoadtestProto.CheckResponse>() {
+      @Override
+      public void onNext(LoadtestProto.CheckResponse checkResponse) {
+        log.debug("Connected to client.");
+        if (checkResponse.getIsFinished()) {
+          clientStatus = ClientStatus.STOPPED;
+          doneFuture.set(null);
+        }
+        if (System.currentTimeMillis() < burnInTimeMillis) {
+          return;
+        }
+        synchronized (this) {
+          for (int i = 0; i < LatencyDistribution.LATENCY_BUCKETS.length; i++) {
+            bucketValues[i] += checkResponse.getBucketValues(i);
+          }
+        }
+      }
+
+      @Override
+      public void onError(Throwable throwable) {
+        if (errors > 3) {
+          clientStatus = ClientStatus.FAILED;
```

Comment:
Which client is this supposed to be the status of? The thing running locally or the load-generating thing in GCE? I think it's the latter, but that's confusing because this file is called Client.java. Also, if it's the latter, I don't think we're differentiating between:

- Failure to reach the server
- The load generation completing successfully
- The load generation completing prematurely with an error.